src/wp-includes/media.php

```diff
@@ -5468,53 +5550,91 @@ function wp_get_webp_info( $filename ) {
- * appear above the fold, in which case it returns a boolean `false`, which will lead to the `loading` attribute being
- * omitted on the element. The purpose of this refinement is to avoid lazy-loading elements that are within the initial
- * viewport, which can have a negative performance impact.
+ * Gets loading optimization attributes.
  *
- * Under the hood, the function uses {@see wp_increase_content_media_count()} every time it is called for an element
- * within the main content. If the element is the very first content element, the `loading` attribute will be omitted.
- * This default threshold of 3 content elements to omit the `loading` attribute for can be customized using the
- * {@see 'wp_omit_loading_attr_threshold'} filter.
- *
- * @since 5.9.0
+ * @since 6.3.0
  *
  * @global WP_Query $wp_query WordPress Query object.
  *
- * @param string $context Context for the element for which the `loading` attribute value is requested.
- * @return string|bool The default `loading` attribute value. Either 'lazy', 'eager', or a boolean `false`, to indicate
- *                     that the `loading` attribute should be skipped.
+ * @param string $tag_name The tag name.
+ * @param array  $attr     Array of the attributes for the tag.
+ * @param string $context  Context for the element for which the loading optimization attribute is requested.
+ * @return array Loading optimization attributes.
  */
-function wp_get_loading_attr_default( $context ) {
+function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 	global $wp_query;
 
-	// Skip lazy-loading for the overall block template, as it is handled more granularly.
+	/*
+	 * Closure for postprocessing logic.
+	 * It is here to avoid duplicate logic in many places below, without having
+	 * to introduce a very specific private global function.
+	 */
+	$postprocess = function( $loading_attributes, $with_fetchpriority = false ) use ( $tag_name, $attr, $context ) {
+		if ( $with_fetchpriority ) {
+			$loading_attributes = wp_maybe_add_fetchpriority_high_attr( $loading_attributes, $tag_name, $attr );
+		}
+		return wp_maybe_strip_loading_attribute( $loading_attributes, $tag_name, $context );
+	};
+
+	$loading_attrs = array();
+
+	/*
+	 * Skip lazy-loading for the overall block template, as it is handled more granularly.
+	 * The skip is also applicable for `fetchpriority`.
+	 */
 	if ( 'template' === $context ) {
-		return false;
+		return $loading_attrs;
+	}
+
+	// For now this function only supports images and iframes.
+	if ( 'img' !== $tag_name && 'iframe' !== $tag_name ) {
+		return $loading_attrs;
+	}
+
+	// For any resources, width and height must be provided, to avoid layout shifts.
+	if ( ! isset( $attr['width'] ) || ! isset( $attr['height'] ) ) {
+		return $loading_attrs;
+	}
+
+	if ( isset( $attr['loading'] ) ) {
+		if ( 'lazy' === $attr['loading'] ) {
+			$loading_attrs['loading'] = $attr['loading'];
```

Comment:
Why are we only copying over this attribute if it is lazy?

Reply:
The main idea of this function is to see if the tag can be lazy-loaded or set high fetchpriority. Even though we respect what user has entered earlier in our logic, we only actually set them in `loading_attrs` if they are either `loading=lazy` or `fetchpriority=high`.

Reply:
@kt-12 What happens to if `loading= eager `, if eager is ignored here, that is unclear why. Maybe a comment in the code.

Reply:
@spacedmonkey Functionally, it doesn't make a difference. `$loading_attrs` would only include `loading="eager"` if it's already present in `$attr`, and then an `array_merge()` wouldn't change anything.

The reason we only add `loading="lazy"` to `$loading_attrs` is consistency. The purpose of this function is to _optimize_ loading attributes and `loading="eager"` is not strictly part of that.

I've expanded the function documentation to clarify: https://github.com/WordPress/wordpress-develop/pull/4495/commits/c061bc4d9cf2e292102b0d9deb31b81a5e5e549d

Reply:
I feel like this is unclear from the code. A comment maybe?

Reply:
@spacedmonkey I added a comment on both of the relevant occurrences in https://github.com/WordPress/wordpress-develop/pull/4495/commits/0c1644afcdda66a4bb6e9457b5946ad3c9f7ee12.